src/Shared/JsonSchemaExporter/JsonSchemaExporter.cs

```diff
@@ -0,0 +1,1128 @@
+        Debug.Assert(schemaType is JsonSchemaType.Integer or JsonSchemaType.Number, "schema type must be number or integer");
+        Debug.Assert(!isIeeeFloatingPoint || schemaType is JsonSchemaType.Number, "If specifying IEEE the schema type must be number");
+
+        string? pattern = null;
+
+        if ((numberHandling & (JsonNumberHandling.AllowReadingFromString | JsonNumberHandling.WriteAsString)) != 0)
+        {
+            pattern = schemaType is JsonSchemaType.Integer
+                ? @"^-?(?:0|[1-9]\d*)$"
+                : isIeeeFloatingPoint
+                    ? @"^-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?$"
+                    : @"^-?(?:0|[1-9]\d*)(?:\.\d+)?$";
+
+            schemaType |= JsonSchemaType.String;
+        }
+
+        if (isIeeeFloatingPoint && (numberHandling & JsonNumberHandling.AllowNamedFloatingPointLiterals) != 0)
+        {
+            return new JsonSchema
+            {
+                AnyOf = new()
+                {
+                    new JsonSchema { Type = schemaType, Pattern = pattern },
+                    new JsonSchema { Enum = new() { (JsonNode)"NaN", (JsonNode)"Infinity", (JsonNode)"-Infinity" } },
+                },
+            };
+        }
+
+        return new JsonSchema { Type = schemaType, Pattern = pattern };
+    }
+
+    // Uses reflection to determine the element type of an enumerable or dictionary type
+    // Workaround for https://github.com/dotnet/runtime/issues/77306#issuecomment-2007887560
+    private static Type GetElementType(JsonTypeInfo typeInfo)
+    {
+        Debug.Assert(typeInfo.Kind is JsonTypeInfoKind.Enumerable or JsonTypeInfoKind.Dictionary, "TypeInfo must be of collection type");
+        _elementTypeProperty ??= typeof(JsonTypeInfo).GetProperty("ElementType", BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
+        return (Type)_elementTypeProperty?.GetValue(typeInfo)!;
+    }
+
+    private static PropertyInfo? _elementTypeProperty;
+
+    // The .NET 8 source generator doesn't populate attribute providers for properties
+    // cf. https://github.com/dotnet/runtime/issues/100095
+    // Work around the issue by running a query for the relevant MemberInfo using the internal MemberName property
+    // https://github.com/dotnet/runtime/blob/de774ff9ee1a2c06663ab35be34b755cd8d29731/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs#L206
+    [RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
+    private static ICustomAttributeProvider? ResolveAttributeProvider(Type? declaringType, JsonPropertyInfo? propertyInfo)
+    {
+        if (declaringType is null || propertyInfo is null)
+        {
+            return null;
+        }
+
+        if (propertyInfo.AttributeProvider is { } provider)
+        {
+            return provider;
+        }
+
+        _memberNameProperty ??= typeof(JsonPropertyInfo).GetProperty("MemberName", BindingFlags.Instance | BindingFlags.NonPublic)!;
```

Comment:
If we haven't already, we should add a test to runtime that does the same operation in the same way and will fail if we ever change it.

Reply:
We should, we originally capped the lifetime of this component to November but I guess these can't ever be changed until net8.0 falls out of support.